File: haskell/glue/src/Glue/Lib/Math/Power/Pow.hs
module Glue.Lib.Math.Power.Pow where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Power function (base^exponent)
-- Mirrors Haskell Glue.Lib.Math.Power.Pow.pow exactly
pow :: IR Eval
pow = NativeFunc powImpl

-- Power function implementation
-- Mirrors Haskell Glue.Lib.Math.Power.Pow.powImpl exactly
powImpl :: IR Eval -> Eval (IR Eval)
powImpl arg1 = pure $ NativeFunc (powWith arg1)

powWith :: IR Eval -> IR Eval -> Eval (IR Eval)
powWith arg1 arg2 = case (arg1, arg2) of
    (Integer n1, Integer n2) -> pure $ Integer (n1 ^ n2)
    (Integer n1, Float n2) -> pure $ Float (fromIntegral n1 ** n2)
    (Float n1, Integer n2) -> pure $ Float (n1 ** fromIntegral n2)
    (Float n1, Float n2) -> pure $ Float (n1 ** n2)
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/math/power/pow.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Power function (base^exponent)
/// Mirrors Haskell Glue.Lib.Math.Power.Pow.pow exactly
final Ir pow = IrNativeFunc(powImpl);

/// Power function implementation (base^exponent)
/// Mirrors Haskell Glue.Lib.Math.Power.Pow.powImpl exactly
Eval<Ir> powImpl(Ir base) {
  return Eval.pure(IrNativeFunc(powTo(base)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Power.Pow.powTo exactly
Eval<Ir> Function(Ir) powTo(Ir base) {
  return (Ir exponent) {
    return sequenceAll([eval(base), eval(exponent)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final b), IrInteger(value: final e)] => Eval.pure(
          IrInteger(math.pow(b, e).toInt()),
        ),
        [IrInteger(value: final b), IrFloat(value: final e)] => Eval.pure(
          IrFloat(math.pow(b.toDouble(), e).toDouble()),
        ),
        [IrFloat(value: final b), IrInteger(value: final e)] => Eval.pure(
          IrFloat(math.pow(b, e.toDouble()).toDouble()),
        ),
        [IrFloat(value: final b), IrFloat(value: final e)] => Eval.pure(
          IrFloat(math.pow(b, e).toDouble()),
        ),
        _ => throwError(wrongArgumentType(['number', 'number'])),
      };
    });
  };
}
